crates/red_knot/src/watch/watcher.rs

```diff
@@ -0,0 +1,393 @@
+                        }
+                    }
+                }
+
+                // No more file changes after 10 ms, send the changes and schedule a new analysis
+                let events = debouncer.into_events();
+
+                if !events.is_empty() {
+                    handler.handle(events);
+                }
+            }
+        })
+        .unwrap();
+
+    let debouncer_sender = sender.clone();
+    let watcher =
+        recommended_watcher(move |event| sender.send(DebouncerMessage::Event(event)).unwrap())?;
+
+    Ok(Watcher {
+        watcher,
+        debouncer_sender,
+        debouncer_thread: Some(debouncer),
+    })
+}
+
+#[derive(Debug)]
+enum DebouncerMessage {
+    /// A new file system event.
+    Event(notify::Result<notify::Event>),
+
+    Flush,
+
+    /// Exit the debouncer thread.
+    Exit,
+}
+
+pub struct Watcher {
+    watcher: RecommendedWatcher,
+    debouncer_sender: crossbeam::channel::Sender<DebouncerMessage>,
+    debouncer_thread: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Watcher {
+    /// Sets up file watching for `path`.
+    pub fn watch(&mut self, path: &SystemPath) -> notify::Result<()> {
+        self.watcher
+            .watch(path.as_std_path(), RecursiveMode::Recursive)
+    }
+
+    /// Stops file watching for `path`.
+    pub fn unwatch(&mut self, path: &SystemPath) -> notify::Result<()> {
+        self.watcher.unwatch(path.as_std_path())
+    }
+
+    /// Stops the file watcher.
+    ///
+    /// Pending events will be discarded.
+    ///
+    /// The call blocks until the watcher has stopped.
+    pub fn stop(mut self) {
+        self.set_stop();
+        if let Some(debouncher) = self.debouncer_thread.take() {
+            debouncher.join().unwrap();
+        }
+    }
+
+    /// Flushes any pending events.
+    pub fn flush(&self) {
+        self.debouncer_sender.send(DebouncerMessage::Flush).unwrap();
+    }
+
+    fn set_stop(&mut self) {
+        self.debouncer_sender.send(DebouncerMessage::Exit).ok();
+    }
+}
+
+impl Drop for Watcher {
+    fn drop(&mut self) {
+        self.set_stop();
+    }
+}
+
+#[derive(Default)]
+struct Debouncer {
+    events: Vec<ChangeEvent>,
+    rescan_event: Option<ChangeEvent>,
+}
+
+impl Debouncer {
+    #[tracing::instrument(level = "trace", skip(self))]
+    fn add_result(&mut self, result: notify::Result<notify::Event>) {
+        match result {
+            Ok(event) => self.add_event(event),
+            Err(error) => self.add_error(error),
+        }
+    }
+
+    #[allow(clippy::unused_self, clippy::needless_pass_by_value)]
+    fn add_error(&mut self, error: notify::Error) {
+        // Micha: I skimmed through some of notify's source code and it seems the most common errors
+        // are IO errors. All other errors should really only happen when adding or removing a watched folders.
+        // It's not clear what an upstream handler should do in the case of an IOError (other than logging it).
```

Comment:
Is this a case where `Rescan` might make sense? Or is that too aggressive?

Reply:
Possibly, not sure. I honestly don't understand when this is happening well enough yet. We may also have to propagate errors to `apply_changes` for it to decide what to do in the case of an error. 

For now, I prefer to leave it as is and wait until we can do some testing on real world projects (e.g. by navigating through individual commits) and hopefully that allows us to trigger some of the errors and then decide what we should do about them.